### ISSUE TITLE
Improve import speed with lazy initialization

### ISSUE THREAD

**Issue by ppetrushkov** [NVIDIA/TransformerEngine#624]
Currently importing transformer_engine takes ~10s on my machine and it also starts a background process pool because of all the JIT initialization like [here](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/pytorch/jit.py#L50-L54) .

It would be better if all the expensive parts were initialized lazily or at least when importing specific subpackage/module, but not when importing root level package.

This also affects e.g. "import accelerate" which uses transformer_engine, which is pretty annoying as I often use accelerate for workloads unrelated to transformer_engine.

**Comment by ptrendx** [comment]
@ksivaman Could you take a look at it? Maybe we could move jitting to the actual modules? In general I agree that we should not do anything CUDA-related on just import, but rather on the first use.

**Comment by afcruzs** [comment]
I've also noted that when importing certain submodules of the pytorch package it slows down so much that vscodes's debugger hangs :( . An example of this is `from transformer_engine.pytorch.float8_tensor import Float8Tensor` -> This induces already a hang for me when training with multiple processes.

**Comment by 980202006** [comment]
I encountered a similar hang problem. I used megatron-lm and after installing transformer-engine, I encountered a hang at torch.distributed.init. This problem will not occur after uninstalling

**Comment by cnut1648** [comment]
Agreed, this takes more than 3 minutes on our server. Is there any way to bypass this issue?

**Comment by ZetangForward** [comment]
@ksivaman Same question, is this question solved currently? Which version of TE should I install to avoid this? Thx

> I encountered a similar hang problem. I used megatron-lm and after installing transformer-engine, I encountered a hang at torch.distributed.init. This problem will not occur after uninstalling. I observed that during the hanging process, many sub processes were created. ![image](https://github.com/NVIDIA/TransformerEngine/assets/24452502/70053310-a4ad-4d46-8892-edc0d6ab0648)

**Comment by JungHoyoun** [comment]
@ZetangForward 
I installed version 2.4.0 and the hang issue has been solved

**Comment by ksivaman** [comment]
Closed by https://github.com/NVIDIA/TransformerEngine/pull/1599
